### ISSUE TITLE
Error in rule mask: Chunk shape must not be greater than data shape in any dimension.

### ISSUE THREAD

**Issue by WeijiaSu** [STOmics/SAW#167]
Hello,

I am running SAW v8.1.1 on two large chips (3x5).
I got an error in the filtering mask file step, it says 

"RuleException:
ValueError in file /saw-8.1.1/config/base.smk, line 64:
Chunk shape must not be greater than data shape in any dimension. (3, 102903, 1) is not compatible with (64680, 102900, 1)"

The second chip had the similar error message: "Chunk shape must not be greater than data shape in any dimension. (3, 61743, 1) is not compatible with (102900, 61740, 1);"

It seems that the chunk size is always greater by 3 in the second dimension. 

I never got this error in the small chips.

Any suggestion would be appreciated. 
Thanks.

**Comment by Jinghao-Chen** [comment]
@WeijiaSu Hi, would you mind providing the chip number?

**Comment by WeijiaSu** [comment]
Hello,

Thanks for your reply. The mask file is C04287A1F3.barcodeToPos.h5

Thanks for your help.


KentLE ***@***.***> 于2024年11月27日周三 11:16写道：

**Comment by Jinghao-Chen** [comment]
@WeijiaSu Hello, I'm sorry that this is due to the limitation of h5py. This problem will be fixed in SAW 8.1.2, which will be released in two days (November 30 GMT+8). If you want to run it with SAW 8.1.1 now, you could temporarily modify line 64 of `saw-8.1.1/config/base.smk` and change`chunks=d.chunks` to `chunks=(1537,1537,1)`

After modification, it is as follows
`dset = fd.create_dataset('bpMatrix_1', data=d[:height, :width], chunks=(1537, 1537, 1), compression=d.compression, compression_opts=d.compression_opts)`

**Comment by WeijiaSu** [comment]
Hello @Clouate ,

Thanks for your reply. I have modified this. But it seems that this method only works for this particular mask. We have many other masks still having this issue after modifying the base.smk file. 
Also, I tried to download the SAW 8.1.2 this morning, but it was not available on the website, the latest version was still SAW 8.1.1 (October 2024).

Thanks,
Weijia

**Comment by Jinghao-Chen** [comment]
@WeijiaSu Sorry for the delay in the release of SAW 8.1.2, I will tell you as soon as it is released. Would you mind providing the chip number that still have problems after you modified the base.smk?

**Comment by WeijiaSu** [comment]
> @WeijiaSu Sorry for the delay in the release of SAW 8.1.2, I will tell you as soon as it is released. Would you mind providing the chip number that still have problems after you modified the base.smk?

hello @Clouate , the chip number is Y00709M7PC; 
error message: Chunk shape must not be greater than data shape in any dimension. (3, 102903, 1) is not compatible with (64680, 102900, 1)

**Comment by Jinghao-Chen** [comment]
@WeijiaSu Hi, I tried to analyze this chip using the modified base.smk, and it worked successfully. And the error message seems to appear before the modification. Could you confirm that you have used the modified base.smk and try again?

**Comment by WeijiaSu** [comment]
Hello @Clouate,

Thank you for your reply. It seems the issue with the second chip was due to a duplicated copy of SAW. I have fixed this and apologize for the inconvenience caused.

For the masks that passed this step after the modification, the alignment proceeded without any issues. However, I encountered an error during the annotation step, and I am unsure if it is related to this issue.

The command executed was:

```
/saw-8.1.1/config/base.smk, line 290:
Exit Snakemake program: The command Error:  
/saw-8.1.1/lib/saw_count/Bam2Gem -i  /SAW_analysis/C04287A1F3/run1/result/C04287A1F3/STEREO_ANALYSIS_WORKFLOW_PROCESSING/ALIGNMENT/fastq1_1.Aligned.sortedByCoord.out.bam, /SAW_analysis/C04287A1F3/run1/result/C04287A1F3/STEREO_ANALYSIS_WORKFLOW_PROCESSING/ALIGNMENT/fastq2_1.Aligned.sortedByCoord.out.bam 
-o  /SAW_analysis/C04287A1F3/run1/result/C04287A1F3/STEREO_ANALYSIS_WORKFLOW_PROCESSING/ANNOTATION/C04287A1F3.Aligned.sortedByCoord.out.merge.q10.dedup.target.bam 
-a  /input_files/references/v8/human_transcriptome/genes/Homo_sapiens.GRCh38.93.gtf 
-s  /SAW_analysis/C04287A1F3/run1/result/C04287A1F3/STEREO_ANALYSIS_WORKFLOW_PROCESSING/ANNOTATION/C04287A1F3.Aligned.sortedByCoord.out.merge.q10.dedup.target.bam.summary.stat 
-e  /SAW_analysis/C04287A1F3/run1/result/C04287A1F3/STEREO_ANALYSIS_WORKFLOW_PROCESSING/ANNOTATION/C04287A1F3.raw.gef 
--sat_file  /SAW_analysis/C04287A1F3/run1/result/C04287A1F3/STEREO_ANALYSIS_WORKFLOW_PROCESSING/ANNOTATION/C04287A1F3_raw_barcode_gene_exp.txt 
--maskSize=61739,102899 --umi_on --umi_len 10 --save_lq --save_dup --snC04287A1F3 -c 24   umi_filter 
--mask  /SAW_analysis/C04287A1F3/run1/result/C04287A1F3/STEREO_ANALYSIS_WORKFLOW_PROCESSING/ALIGNMENT/C04287A1F3.barcodeToPos.h5 
> /SAW_analysis/C04287A1F3/run1/result/C04287A1F3/STEREO_ANALYSIS_WORKFLOW_PROCESSING/ANNOTATION/log_anno
```

The error message in the ANNOTATION/log_anno:

>[2024-11-29 06:20:26] ----start annotation------
>[2024-11-29 06:20:31] chr cnt:47 genecnt:58395
>[2024-11-29 06:20:31] loadgtf - elapsed time: 4144.66732 ms
>[2024-11-29 06:20:31] getchrname - elapsed time: 51.62480 ms
>[2024-11-29 06:20:34] Could not open record file.

I have verified that all the input files, including the human reference and the GTF files, are accessible and functional. These references have been successfully used multiple times before.

Could you please clarify what the "record file" mentioned in the log_anno file refers to?

Thank you for your help.

Best regards

**Comment by Jinghao-Chen** [comment]
@WeijiaSu Hi, the record file is the intermediate file for multi-threaded execution in annotation step, the reason for this error may be the number of files in your ANNOTATION directory has exceeded the limit of your linux system. You could delete all files in the ANNOTATION directory, including hidden files, and then rerun SAW with the same command. In addition, you may need to set more file number limits, such as `ulimit -n 204800`

**Comment by Jinghao-Chen** [comment]
@WeijiaSu Hi, `SAW 8.1.2` is now available on the website. https://stereotoolss-organization.gitbook.io/saw-user-manual-v8.1/download-center#saw-8.1.2-nov.-2024
 Sorry again for the inconvenience.

Best regards

**Comment by WeijiaSu** [comment]
Hi @Clouate ,

Thank you for the update. We are currently using SAW 8.1.2. The CID mapping and alignment steps completed successfully. 
However, the program appears to be stuck at the annotation step.

It generated a merged.bam file of approximately 250 GB, whereas the total BAM files in the ALIGNMENT directory are around 900 GB. The SAW program is still running but has not been generating any new output for over 20 hours. The sample is a 3*5 chip.

Should I terminate the process and re-run it? If so, would Snakemake resume from the annotation step?

Thank you for your assistance!

**Comment by Jinghao-Chen** [comment]
Hi, @WeijiaSu ,could you check if the hidden files have been updated by `ls -alht`? Or, could you provide ANNOTATION/anno*log? In addition, there is a bug in SAW, that is, when the gtf file contains extra chromosomes that do not belong to the header of ALIGNMENT/bam(`samtools view -H ALIGNMENT/bam`), the program will get stuck. You could also confirm this. Of course, the processing time of large chip data will be very long because the data is really large.

**Comment by WeijiaSu** [comment]
Hi @Clouate 
Thanks for your reply. I have checked the progress based on your suggestion:

1. "ls -alht" shows the hidden files, the only hidden file that's been updating in real time is ".target.bam", which is 188G and SLOWLY increasing.
2. I have checked the headers in ALIGMENT/bam and gft, there was no extra chromosome in gtf that do not belong to the bam  headers.

With this, do you think we should just wait the program to run due to the large data for the big chips?

Thanks

**Comment by Jinghao-Chen** [comment]
@WeijiaSu Yes, target.bam is being updated, indicating that the annotation program is running the final adjustment, which is quite time-consuming even on a 1*1 chip, so you could patiently wait for the execution to complete.
In addition, using solid state drives can speed up processing.
